### DISCUSSION TITLE
What is the best way to print out a custom attribute added on a middleware after Sidekiq 6.4.1

### DISCUSSION THREAD

**Discussion by wsurapat-mdsol** [sidekiq/sidekiq#5343] [Q&A] [answered]
Hi, I would appreciate if anyone can guide me how to print out a custom attribute added on a middleware.
I had this middleware written before Sidekiq 6.4.1.
```ruby
module SidekiqMiddleware
  module Client
    class Hello
      def call(_worker_class, job, _queue, _redis_pool)
        job["hello"] = "world"

        yield
      end
    end
  end
  
  ....
end
  ```
I registered it together with a custom logger.
```ruby
Sidekiq.configure_server do |config|
  config.client_middleware do |chain|
    chain.add SidekiqMiddleware::Client::Hello
  end

  config.server_middleware do |chain|
    chain.add SidekiqMiddleware::Server::Hello
  end

  config.log_formatter = Sidekiq::Logger::Formatters::JSON.new
  config.options[:job_logger] = SidekiqJobLoggerWithHello

  SidekiqUniqueJobs::Server.configure(config)
end
```
And the custom logger looks like this.
```ruby
require "sidekiq/job_logger"

class SidekiqJobLoggerWithHello < Sidekiq::JobLogger
  def job_hash_context(job_hash)
    original_job_hash = super
    original_job_hash.merge("hello" => job_hash["hello"])
  end
end
```

This code works before, but after upgrading to 6.4.2, it didn't work. I saw that there is this change https://github.com/mperham/sidekiq/blob/00ae6a74234a9b626a593ae6c5620416349f8eaf/lib/sidekiq/job_logger.rb#L24 introduced in 6.4.1 which removed `job_hash_context` method. Thus the code I have didn't work any more and I can not figure out a good way to fix it. Can someone suggest what is the best practice to achieve this?

**Comment by mperham** [comment, marked as answer]
You don't need a custom job logger at all. In your server middleware, you'd just add:

```ruby
Sidekiq::Context.with("hello" => job["hello"], &block)
```

**Comment by wsurapat-mdsol** [comment]
Hi @mperham  thanks for the suggestion. Really appreciated. 
I tried using the Sidekiq::Context as you suggested
```ruby
module SidekiqMiddleware
  module Client
    class Hello
      def call(_worker_class, job, _queue, _redis_pool)
        job["hello"] = "world"

        yield
      end
    end
  end

  module Server
    class Hello
      def call(worker, job, _queue)

        Sidekiq::Context.with("hello" => "world") do
          Sidekiq.logger.info
        end

        yield
      end
    end
  end
end
```

Then, when i trigger the job, I can see the log as follows.
```
{"ts":"2022-05-24T18:09:16.268Z","pid":17955,"tid":"2o3","lvl":"INFO","msg":"start","ctx":{"class":"SillyWorker","jid":"eeaa4854e04736e721f8011e"}}

{"ts":"2022-05-24T18:09:16.289Z","pid":17955,"tid":"2o3","lvl":"INFO","msg":null,"ctx":{"class":"SillyWorker","jid":"eeaa4854e04736e721f8011e","hello":"world"}}

SillyWorker starts
SillyWorker finishes

{"ts":"2022-05-24T18:09:16.289Z","pid":17955,"tid":"2o3","lvl":"INFO","msg":"done","ctx":{"class":"SillyWorker","jid":"eeaa4854e04736e721f8011e","elapsed":0.021}}
```
I can see that the `"hello":"world"` was printed out on the second log that I print out from the block I supplied to `Sidekiq::Context.with`.  I saw the implementation of `.with` from https://github.com/mperham/sidekiq/blob/main/lib/sidekiq/logger.rb#L8, which looks to me like the context that I injected should be preserved. However, I could not see it in the last log when the job is done as shown above. Did I understand something wrong? How can I make the context that I inject preserved until the end?

**Reply by mperham** [reply]
You need to pass the middleware's block. Remove your yield and pass it to with.

**Reply by wsurapat-mdsol** [reply]
Got it, thanks a lot for your help @mperham 🙇
